src/librbd/cache/pwl/AbstractWriteLog.cc

```diff
@@ -1463,8 +1463,8 @@ void AbstractWriteLog<I>::alloc_and_dispatch_io_req(C_BlockIORequestT *req)
 template <typename I>
 bool AbstractWriteLog<I>::check_allocation(
     C_BlockIORequestT *req, uint64_t bytes_cached, uint64_t bytes_dirtied,
-    uint64_t bytes_allocated, uint32_t num_lanes, uint32_t num_log_entries,
-    uint32_t num_unpublished_reserves) {
+    uint64_t bytes_allocated, uint64_t num_lanes, uint64_t num_log_entries,
+    uint64_t num_unpublished_reserves) {
```

Comment:
these values are `rwl` specific, so they can be 32bit since the log entries are limited to a million

Reply:
> these values are `rwl` specific, so they can be 32bit since the log entries are limited to a million

The types here are so confusing that I think it's necessary to make an alignment. Its statement is `uint64_t`, like `uint64_t num_lanes = 0;`. The parameter type passed func is `uint32_t`, like `uint32_t num_lanes`, but the variables involved in the operation are `uint64_t` and `unsigned int`, like 
```
      m_free_lanes -= num_lanes;
      m_free_log_entries -= num_log_entries;
```
`m_free_lanes` is `unsigned int`.
 `m_free_log_entries` is `uint32_t`, it must be change to `uint64_t`.

Except some temporary variables, I think we should not use the compiler dependent types like `unsign int`, `int`, `long` and so on. change these to `uint64_t` and `uint32_t`.
@idryomov What's your opinion? ? I'll make a further revision according to your opinion or leave it as it is.

Reply:
`num_lanes`, `num_log_entries` and `num_unpublished_reserves` parameters are compared with `m_free_lanes`, `m_free_log_entries` and `m_unpublished_reserves` all of which are specific to `rwl` mode and are either `uint32_t` or `unsigned int`.  

This PR is supposed to address issues with >4G cache size in `ssd` mode.  Converting some fields from `unsigned int` to `uint32_t` is not a bad idea but should be done separately.  Let's keep this PR as small as possible so that it is clear what exactly was leading to data corruption at 4G boundary.

Reply:
>  `m_free_log_entries` is `uint32_t`, it must be change to `uint64_t`.

No, `m_free_log_entries` is fine as `uint32_t`.  It is always zero in `ssd` mode and per @MahatiC is limited to a million in `rwl` mode.

The way how some of log entry accounting is shared through `AbstractWriteLog` even though log entries in `ssd` mode are very different from log entries in `rwl` mode is super confusing but again it wasn't me who designed this.